File: LuaJIT/Chords.lua
local math = math
local abs = math.abs
local huge = math.huge

return function(f, a, b, eps, maxIter)
    eps = eps or 1e-12
    maxIter = maxIter or 1e8

    local fa = f(a)
    local fb = f(b)

    if fa * fb > 0 then
        return huge
    end

    local c = a

    for i = 1, maxIter, 1 do
        c = (a * fb - b * fa) / (fb - fa)
        local fc = f(c)

        if abs(fc) < eps then
            return c
        end

        if fa * fc < 0 then
            b = c
            fb = fc
        else
            a = c
            fa = fc
        end
    end

    return huge
end

File: Python-Numba/chords.py
from typing import Callable
from numba import jit
import math


range_f = range
abs_f = abs
inf = math.inf


@jit
def chord_method(func: Callable[[float], float], a: float, b: float, eps: float = 1e-12, max_iter: int = 10**8):
    if func(a) * func(b) >= 0:
        return inf

    x_prev = a
    x_curr = b

    for i in range_f(max_iter):
        x_next = x_curr - func(x_curr) * (x_curr - x_prev) / (func(x_curr) - func(x_prev))

        if abs_f(x_next - x_curr) < eps:
            return x_next

        x_prev, x_curr = x_curr, x_next

    return x_curr
